File: backend/app/routers/networks.py
"""Network management endpoints."""
from __future__ import annotations

from docker.errors import APIError, NotFound
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from ..auth import User, authenticate, require_admin
from ..docker_client import get_client

router = APIRouter(prefix="/api/networks", tags=["networks"])


class CreateNetwork(BaseModel):
    name: str
    driver: str = "bridge"
    internal: bool = False
    attachable: bool = True
    labels: dict[str, str] | None = None


def _summary(n) -> dict:
    attrs = n.attrs or {}
    return {
        "id": n.id,
        "short_id": n.short_id,
        "name": n.name,
        "driver": attrs.get("Driver"),
        "scope": attrs.get("Scope"),
        "internal": attrs.get("Internal"),
        "attachable": attrs.get("Attachable"),
        "ipam": attrs.get("IPAM"),
        "labels": attrs.get("Labels") or {},
        "containers": list((attrs.get("Containers") or {}).keys()),
    }


@router.get("")
def list_networks(_: User = Depends(authenticate)) -> list[dict]:
    try:
        return [_summary(n) for n in get_client().networks.list()]
    except APIError as exc:
        raise HTTPException(status_code=502, detail=str(exc))


@router.get("/{network_id}")
def inspect_network(network_id: str, _: User = Depends(authenticate)) -> dict:
    try:
        return get_client().networks.get(network_id).attrs
    except NotFound:
        raise HTTPException(status_code=404, detail="Network not found")
    except APIError as exc:
        raise HTTPException(status_code=502, detail=str(exc))


@router.post("")
def create(req: CreateNetwork, _: User = Depends(require_admin)) -> dict:
    try:
        n = get_client().networks.create(
            name=req.name,
            driver=req.driver,
            internal=req.internal,
            attachable=req.attachable,
            labels=req.labels or {},
        )
        return _summary(n)
    except APIError as exc:
        raise HTTPException(status_code=502, detail=str(exc))


@router.delete("/{network_id}")
def remove(network_id: str, _: User = Depends(require_admin)) -> dict:
    try:
        get_client().networks.get(network_id).remove()
        return {"removed": network_id}
    except NotFound:
        raise HTTPException(status_code=404, detail="Network not found")
    except APIError as exc:
        raise HTTPException(status_code=502, detail=str(exc))


@router.post("/prune")
def prune(_: User = Depends(require_admin)) -> dict:
    try:
        return get_client().networks.prune()
    except APIError as exc:
        raise HTTPException(status_code=502, detail=str(exc))

File: backend-node/src/routes/networks.js
// Network management endpoints.
import { Type } from '@sinclair/typebox';
import { getClient } from '../docker-client.js';
import { asyncHandler } from '../util.js';
import { createApiRouter } from '../route-builder.js';
import {
  ConnectRequest,
  CreateNetworkRequest,
  DisconnectRequest,
  NetworkSummary,
  PassThroughObject,
} from '../schemas/index.js';

const r = createApiRouter('/api/networks', { tag: 'networks' });

function summary(n) {
  return {
    id: n.Id,
    short_id: (n.Id || '').slice(0, 12),
    name: n.Name,
    driver: n.Driver,
    scope: n.Scope,
    internal: n.Internal,
    attachable: n.Attachable,
    ipam: n.IPAM,
    labels: n.Labels || {},
    containers: Object.keys(n.Containers || {}),
  };
}

const IdParam = Type.Object({ id: Type.String() }, { additionalProperties: false });

r.get(
  '/',
  { summary: 'List networks', responses: { 200: Type.Array(NetworkSummary) } },
  asyncHandler(async (_req, res) =>
    res.json((await getClient().listNetworks()).map(summary)),
  ),
);

r.post(
  '/',
  {
    summary: 'Create a network',
    admin: true,
    destructive: true,
    body: CreateNetworkRequest,
    responses: { 200: NetworkSummary },
  },
  asyncHandler(async (req, res) => {
    const b = req.body;
    const n = await getClient().createNetwork({
      Name: b.name,
      Driver: b.driver,
      Internal: b.internal,
      Attachable: b.attachable,
      Labels: b.labels || {},
    });
    res.json(summary(await n.inspect()));
  }),
);

r.post(
  '/prune',
  { summary: 'Prune unused networks', admin: true, destructive: true, responses: { 200: PassThroughObject } },
  asyncHandler(async (_req, res) => res.json(await getClient().pruneNetworks())),
);

r.get(
  '/:id',
  { summary: 'Inspect a network', params: IdParam, responses: { 200: PassThroughObject } },
  asyncHandler(async (req, res) =>
    res.json(await getClient().getNetwork(req.params.id).inspect()),
  ),
);

r.delete(
  '/:id',
  {
    summary: 'Remove a network',
    admin: true,
    destructive: true,
    params: IdParam,
    responses: { 200: Type.Object({ removed: Type.String() }) },
  },
  asyncHandler(async (req, res) => {
    await getClient().getNetwork(req.params.id).remove();
    res.json({ removed: req.params.id });
  }),
);

r.post(
  '/:id/connect',
  {
    summary: 'Connect a container to a network',
    admin: true,
    destructive: true,
    params: IdParam,
    body: ConnectRequest,
    responses: { 200: PassThroughObject },
  },
  asyncHandler(async (req, res) => {
    const b = req.body;
    await getClient().getNetwork(req.params.id).connect({
      Container: b.container,
      EndpointConfig: {
        Aliases: b.aliases || undefined,
        Links: b.links || undefined,
        IPAMConfig:
          b.ipv4_address || b.ipv6_address
            ? { IPv4Address: b.ipv4_address || undefined, IPv6Address: b.ipv6_address || undefined }
            : undefined,
      },
    });
    res.json({ network: req.params.id, container: b.container, connected: true });
  }),
);

r.post(
  '/:id/disconnect',
  {
    summary: 'Disconnect a container from a network',
    admin: true,
    destructive: true,
    params: IdParam,
    body: DisconnectRequest,
    responses: { 200: PassThroughObject },
  },
  asyncHandler(async (req, res) => {
    const b = req.body;
    await getClient().getNetwork(req.params.id).disconnect({
      Container: b.container,
      Force: b.force,
    });
    res.json({ network: req.params.id, container: b.container, disconnected: true });
  }),
);

export default r;
